### PR TITLE
SIL optimizer: remove unbalanced retains/releases from immortal objects

### PR DESCRIPTION
**Pull request by eeckstein**
ARC operations don't have an effect on immortal objects, like the empty array singleton or statically allocated arrays.
Therefore we can freely remove and retain/release instructions on such objects, even if there is no paired balanced ARC operation.

This optimization can only be done with a minimum deployment target of Swift 5.1, because in that version we added immortal ref count bits.

The optimization is implemented in libswift. Additionally, the remaining logic of simplifying strong_retain and strong_release is also ported to libswift.

This PR also includes some infrastructure improvements in libswift. For details see the commit list.

rdar://81482156

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

File: libswift/Sources/Optimizer/InstructionPasses/SimplifyStrongRetainRelease.swift [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  switch value {
+    case let cfi as ConvertFunctionInst:
+      return isNotReferenceCounted(value: cfi.operand, context: context)
+    case let uci as UpcastInst:
+      return isNotReferenceCounted(value: uci.operand, context: context)
+    case let urc as UncheckedRefCastInst:
+      return isNotReferenceCounted(value: urc.operand, context: context)
+    case is GlobalValueInst:
+      // Since Swift 5.1, statically allocated objects have "immortal" reference
+      // counts. Therefore we can savely eliminate unbalaced retains and
```
Comment (cltnschlosser) [inline libswift/Sources/Optimizer/InstructionPasses/SimplifyStrongRetainRelease.swift:81]:
```suggestion
      // counts. Therefore we can safely eliminate unbalaced retains and
```

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci test macOS

**Pull request merged**